### PR TITLE
Container timeout only removes local cache in session replication.

### LINKED ISSUE #3362: hazelcast web session replication doesn't replicate when shutdown
full story is here: https://groups.google.com/forum/#!topic/hazelcast/H1FWmMgBdz4
when server goes to shutdown, our architecture couldn't get whether this process is shutdown or timeout.
so in some cases replication doesn't work.

```
    @Test
    public void test() throws Exception {
        CookieStore cookieStore = new BasicCookieStore();
        executeRequest("write", serverPort1, cookieStore);
        server1.restart();
        assertEquals("value",executeRequest("read", serverPort2, cookieStore));
    }
```

### PR DESCRIPTION
**Pull request by kobalski**

### CODE REVIEW

File: hazelcast-wm/src/main/java/com/hazelcast/web/ClusteredSessionService.java [hunk 1]
```diff
@@ -145,6 +151,15 @@ private void reconnectHZInstance() throws ServletException {
         hazelcastInstance = HazelcastInstanceLoader.createInstance(this, filterConfig, properties);
         clusterMap = hazelcastInstance.getMap(clusterMapName);
         sss = (SerializationServiceSupport) hazelcastInstance;
+        try {
+            Config hzConfig = hazelcastInstance.getConfig();
+            MapConfig mapConfig = hzConfig.getMapConfig(clusterMapName);
+            mapConfig.setMaxIdleSeconds(Integer.parseInt(sessionTTL));
+            hzConfig.addMapConfig(mapConfig);
+
```
Comment (mesutcelik) [inline hazelcast-wm/src/main/java/com/hazelcast/web/ClusteredSessionService.java]: config should be created before instance creation.

File: hazelcast-wm/src/main/java/com/hazelcast/web/ClusteredSessionService.java [hunk 2]
```diff
@@ -145,6 +151,15 @@ private void reconnectHZInstance() throws ServletException {
...
+        } catch (UnsupportedOperationException ignored) {
+            LOGGER.info("Client cannot access config.");
```
Comment (mesutcelik) [inline hazelcast-wm/src/main/java/com/hazelcast/web/ClusteredSessionService.java]: I would suggest to create a fail fast check so if `session-ttl-seconds` are set and mode is client/server, we should throw  InvalidConfigException.

File: hazelcast-wm/src/main/java/com/hazelcast/web/entryprocessor/DeleteSessionEntryProcessor.java [hunk 3]
```diff
@@ -74,7 +74,7 @@ public Object process(Map.Entry<String, SessionState> entry) {
             return Boolean.FALSE;
         }
         sessionState.removeJvmId(jvmId);
```
Comment (mesutcelik) [inline hazelcast-wm/src/main/java/com/hazelcast/web/entryprocessor/DeleteSessionEntryProcessor.java]: are we using jvmIds? If not, we can cleanup this on all EPs.

File: hazelcast-wm/src/test/webapp/WEB-INF/hazelcast_with_max_idle.xml [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            <!-- salt value to use when generating the secret key -->
+            <salt>thesalt</salt>
+            <!-- pass phrase to use when generating the secret key -->
+            <password>thepass</password>
+            <!-- iteration count to use when generating the secret key -->
+            <iteration-count>19</iteration-count>
+        </symmetric-encryption>
+    </network>
+    <partition-group enabled="false"/>
+    <map name="default">
```
Comment (mesutcelik) [inline hazelcast-wm/src/test/webapp/WEB-INF/hazelcast_with_max_idle.xml]: Do we really need that many config definitions? We only need to configure default map configuration.

**Comment by mesutcelik** [comment]
I see that PR is fixing that issue as well. https://github.com/hazelcast/hazelcast/issues/2377

**Comment by mesutcelik** [comment]
:+1:

File: hazelcast-wm/src/main/java/com/hazelcast/web/HazelcastInstanceLoader.java [hunk 5]
```diff
@@ -82,7 +96,13 @@ public static HazelcastInstance createInstance(final ClusteredSessionService ses
                 throw new ServletException(e);
             }
         }
-        return createHazelcastInstance(sessionService, instanceName, config);
+        if (sessionTTLConfig == null) {
+            sessionTTLConfig = SESSION_TTL_DEFAULT_SECONDS;
+        }
+        MapConfig mapConfig = config.getMapConfig(mapName);
+        mapConfig.setMaxIdleSeconds(Integer.parseInt(sessionTTLConfig));
+        config.addMapConfig(mapConfig);
```
Comment (bilalyasar) [inline hazelcast-wm/src/main/java/com/hazelcast/web/HazelcastInstanceLoader.java]: do we need to `addMapConfig` again?

File: hazelcast-wm/src/main/java/com/hazelcast/web/HazelcastInstanceLoader.java [hunk 6]
```diff
@@ -82,7 +96,13 @@ public static HazelcastInstance createInstance(final ClusteredSessionService ses
                 throw new ServletException(e);
             }
         }
-        return createHazelcastInstance(sessionService, instanceName, config);
+        if (sessionTTLConfig == null) {
+            sessionTTLConfig = SESSION_TTL_DEFAULT_SECONDS;
+        }
+        MapConfig mapConfig = config.getMapConfig(mapName);
+        mapConfig.setMaxIdleSeconds(Integer.parseInt(sessionTTLConfig));
```
Comment (bilalyasar) [inline hazelcast-wm/src/main/java/com/hazelcast/web/HazelcastInstanceLoader.java]: what if it is not integer, so fail-fast?

**Comment by bilalyasar** [comment]
:+1:

**Comment by kobalski** [comment]
run-lab-run

**Comment by kobalski** [comment]
run-lan-run

**Pull request merged**